cmd/codegen/generator/go/templates/modules.go

```diff
@@ -770,18 +907,61 @@ func (ps *parseState) isOptionalWrapper(typ types.Type) (*types.Named, bool) {
 // This is essentially for helping resolve the pointeriness of types: a type of
 // **T and a variable p becomes T and &&p. This means we can *always* construct
 // an Opts object and unmarshal into it without having nil dereferences.
-func findOptsAccessPattern(t types.Type, access *Statement) (types.Type, *Statement) {
+func (ps *parseState) findOptsAccessPattern(t types.Type, access *Statement) (types.Type, *Statement, bool, error) {
 	switch t := t.(type) {
 	case *types.Pointer:
 		// taking the address of an address isn't allowed - so we use a ptr
 		// helper function
-		t2, val := findOptsAccessPattern(t.Elem(), access)
-		return t2, Id("ptr").Call(val)
-	// case *types.Slice:
-	// 	t2, val := findOptsAccessPattern(t.Elem(), access)
-	// 	return t2, Index().Id(renderNameOrStruct(t.Elem())).Values(val)
+		t2, access2, ok, err := ps.findOptsAccessPattern(t.Elem(), access)
+		if err != nil {
+			return nil, nil, false, err
+		}
+		if ok {
+			return t2, Id("ptr").Call(access2), true, nil
+		}
+		return nil, nil, false, nil
+	case *types.Named:
+		wrappedType, isOptionalType, err := ps.isOptionalWrapper(t)
+		if err != nil {
+			return nil, nil, false, err
+		}
+		if isOptionalType {
+			wrappedNamed, ok := wrappedType.(*types.Named)
+			if !ok {
+				return t, access, true, nil
+			}
+			_, ok = wrappedNamed.Underlying().(*types.Interface)
+			if !ok {
+				return t, access, true, nil
+			}
+			return t, Id("convertOptionalVal").Call(
+				access,
+				Parens(Op("*").Id(formatIfaceImplName(wrappedNamed.Obj().Name()))).Dot("toIface"),
+			), true, nil
+		}
+
+		if _, ok := t.Underlying().(*types.Interface); ok {
+			return t, access.Dot("toIface").Call(), true, nil
+		}
+		return nil, nil, false, nil
+	case *types.Slice:
```

Comment:
Just for my understanding: this is so that we can do covariance right? We need to take a concrete struct type and lift it to an interface type. But we also need to do this for slices of concrete to slices of interfaces?

A few comments in this function (or more examples in the docstring) could be helpful here. As an aside - if you know of a "formal" term for what I've just coined "access patterns" then that would be wonderful :smile:

Reply:
> We need to take a concrete struct type and lift it to an interface type. But we also need to do this for slices of concrete to slices of interfaces?

Yeah exactly, here's an example chunk of code created by this if it's helpful:
```go
case "WithIfaceList":
  var parent Test
  err = json.Unmarshal(parentJSON, &parent)
  if err != nil {
  	fmt.Println(err.Error())
  	os.Exit(2)
  }
  var ifaces []*customIfaceImpl
  if inputArgs["ifaces"] != nil {
  	err = json.Unmarshal([]byte(inputArgs["ifaces"]), &ifaces)
  	if err != nil {
  		fmt.Println(err.Error())
  		os.Exit(2)
  	}
  }
  return (*Test).WithIfaceList(&parent, convertSlice(ifaces, (*customIfaceImpl).toIface)), nil
```

Trying to get everything to work *just right* with Go's json ser/deser while also avoiding things like typed nils was a real headache and the reason why I added what might seem like an excessive number of test cases 😅

> A few comments in this function (or more examples in the docstring) could be helpful here. 

Yes for sure, will do.

> As an aside - if you know of a "formal" term for what I've just coined "access patterns" then that would be wonderful 😄

Honestly I don't mind the name at all 🙂 At least, I can't think of anything better, so unless that changes I think we can stick with it.